pkg/projects/resource_project_role.go

```diff
@@ -0,0 +1,260 @@
+package projects
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/go-resty/resty/v2"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+const customRoleType = "CUSTOM"
+
+var validRoleEnvironments = []string{
+	"DEV",
+	"PROD",
+}
+
+var validRoleActions = []string{
+	"READ_REPOSITORY",
+	"ANNOTATE_REPOSITORY",
+	"DEPLOY_CACHE_REPOSITORY",
+	"DELETE_OVERWRITE_REPOSITORY",
+	"MANAGE_XRAY_MD_REPOSITORY",
+	"READ_RELEASE_BUNDLE",
+	"ANNOTATE_RELEASE_BUNDLE",
+	"CREATE_RELEASE_BUNDLE",
+	"DISTRIBUTE_RELEASE_BUNDLE",
+	"DELETE_RELEASE_BUNDLE",
+	"MANAGE_XRAY_MD_RELEASE_BUNDLE",
+	"READ_BUILD",
+	"ANNOTATE_BUILD",
+	"DEPLOY_BUILD",
+	"DELETE_BUILD",
+	"MANAGE_XRAY_MD_BUILD",
+	"READ_SOURCES_PIPELINE",
+	"TRIGGER_PIPELINE",
+	"READ_INTEGRATIONS_PIPELINE",
+	"READ_POOLS_PIPELINE",
+	"MANAGE_INTEGRATIONS_PIPELINE",
+	"MANAGE_SOURCES_PIPELINE",
+	"MANAGE_POOLS_PIPELINE",
+	"TRIGGER_SECURITY",
+	"ISSUES_SECURITY",
+	"LICENCES_SECURITY",
+	"REPORTS_SECURITY",
+	"WATCHES_SECURITY",
+	"POLICIES_SECURITY",
+	"RULES_SECURITY",
+	"MANAGE_MEMBERS",
+	"MANAGE_RESOURCES",
+}
+
+type Role struct {
+	Name         string   `json:"name"`
+	Description  string   `json:"description"`
+	Type         string   `json:"type"`
+	Environments []string `json:"environments"`
+	Actions      []string `json:"actions"`
+}
+
+func (r Role) Id() string {
+	return r.Name
+}
+
+func (a Role) Equals(b Identifiable) bool {
+	return a.Id() == b.Id()
+}
+
+func rolesToEquatables(roles []Role) []Equatable {
+	var equatables []Equatable
+
+	for _, role := range roles {
+		equatables = append(equatables, role)
+	}
+
+	return equatables
+}
+
+func equatablesToRoles(equatables []Equatable) []Role {
```

Comment:
God, I thought we were past this shit when Java 8 came out 10 years ago

Reply:
Right?! 😠